mne_sandbox/connectivity/cfc.py

```diff
@@ -0,0 +1,623 @@
+
+        # Data is either Raw or Epochs
+        pbar = ProgressBar(n_ep)
+        for itime, (i_tmin, i_tmax) in enumerate(zip(tmin, tmax)):
+            # Pull times of interest
+            with warnings.catch_warnings():  # To suppress a depracation
+                warnings.simplefilter("ignore")
+                # Not sure how to do this w/o copying
+                i_t_data_am = i_f_data_am.copy().crop(i_tmin, i_tmax)
+                i_t_data_ph = i_f_data_ph.copy().crop(i_tmin, i_tmax)
+
+            if concat_epochs is True:
+                # Iterate through each event type and hstack
+                con_data_ph = []
+                con_data_am = []
+                for i_ev in i_t_data_am.event_id.keys():
+                    con_data_ph.append(np.hstack(i_t_data_ph[i_ev]._data))
+                    con_data_am.append(np.hstack(i_t_data_am[i_ev]._data))
+                i_t_data_ph = np.vstack(con_data_ph)
+                i_t_data_am = np.vstack(con_data_am)
+            else:
+                # Just pull all epochs separately
+                i_t_data_ph = i_t_data_ph._data
+                i_t_data_am = i_t_data_am._data
+            # Now make sure that inputs to the loop are ep x chan x time
+            if i_t_data_am.ndim == 2:
+                i_t_data_ph = i_t_data_ph[np.newaxis, ...]
+                i_t_data_am = i_t_data_am[np.newaxis, ...]
+            # Loop through epochs (or epoch grps), each index pair, and funcs
+            data_iter = zip(i_t_data_ph, i_t_data_am)
+            for iep, (ep_ph, ep_am) in enumerate(data_iter):
+                for iix, (i_ix_ph, i_ix_am) in enumerate(ixs_new):
+                    for ix_func, i_pac_func in enumerate(pac_func):
+                        func = getattr(ppac, i_pac_func)
+                        pac[ix_func, iep, iix, i_f_pair, itime] = func(
+                            ep_ph[i_ix_ph], ep_am[i_ix_am],
+                            f_phase, f_amp, filterfn=False)
+            pbar.update_with_increment_value(1)
+    if pac.shape[0] == 1:
+        pac = pac[0]
+    if return_data:
+        return pac, freq_pac, data_ph, data_am
+    else:
+        return pac, freq_pac
+
+
+def _raw_to_epochs_mne(raw, events, tmin, tmax):
+    """Convert Raw data to Epochs w/ some time checks."""
+    events = np.atleast_1d(events)
+    if events.ndim == 1:
+        events = np.vstack([events, np.zeros_like(events),
+                            np.ones_like(events)]).T
+    if events.ndim != 2:
+        raise ValueError('events have incorrect number of dimensions')
+    if events.shape[-1] != 3:
+        raise ValueError('events have incorrect number of columns')
+    # Convert to Epochs using the event times
+    tmin_all = np.min(tmin)
+    tmax_all = np.max(tmax) + (1. / raw.info['sfreq'])
+    return mne.Epochs(raw, events, tmin=tmin_all, tmax=tmax_all, preload=True,
+                      baseline=None)
+
+
+def _pre_filter_ph_am(data, sfreq, ixs, f_ph, f_am, n_cycles=3,
```

Comment:
actually, what if we kept an `n_cycles` parameter, but allowed the user to supply a function as well. The function would be passed `fmin`, `fmax` and `sfreq`, and would return a single float value corresponding to the number of cycles to include. That way people could define whatever windowing functions they want and we'd keep the same number of parameters. We could include a helper function to calculate a scaling bandwidth based on frequency and include it in the docs. WDYT?